Curation api - changes by @ostolop and @rpetry

rpetry opened the pull request:
Could someone please review?

ostolop committed:
Added first version of Curators API, to add properties/ontology terms to assays.

ostolop committed:
Work in progress - experiment loading :)

ostolop committed:
Work in progress - experiment loading ... (part 2) :)

ostolop committed:
Work in progress - experiment loading ... (part 3) :)

ostolop committed:
Work in progress - experiment loading ... (part 4) :) - work in progress not quite working.

ostolop committed:
Work in progress - experiment loading ... (part 5) :) - now with transactions.

rpetry committed:
Ticket #3022  - Changes for implementing Curator's API

alf239 commented on atlas-web/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -473,7 +482,7 @@
 
     <filter-mapping>
-        <filter-name>OpenSessionInViewFilter</filter-name>
+        <filter-name>ReadWriteHibernateSessionFilter</filter-name>
```
We replaced the filter mapping, but did not remove the older filter. Is it as intended?

alf239 commented on atlas-dao/src/main/resources/uk/ac/ebi/gxa/dao/daoContext.xml:
```diff
@@ -1,7 +1,13 @@
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd">
+       xmlns:tx="http://www.springframework.org/schema/tx"
+       xmlns:aop="http://www.springframework.org/schema/aop"
```
AOP? Why?

alf239 commented:
Does not compile.

rpetry committed:
Ticket #3022  - Changes for implementing Curator's API

rpetry committed:
Merge branch 'develop' into curation-api

rpetry committed:
Added testAddDeleteAssayProperty() test

alf239 committed:
Finishing removal of releaseDate field so that code would compile

alf239 committed:
Adjusting HSQL sequences so that they would not clash with the existing data

alf239 committed:
Addressing `org.hibernate.AssertionFailure: no collection snapshot for orphan delete`

  - `delete-orphan` does not work with immutable collections (see e.g. https://forum.hibernate.org/viewtopic.php?p=2376362)

alf239 committed:
Not @Immutable any more

alf239 committed:
Code style

alf239 committed:
Proper association (ID removed)

alf239 committed:
Rephrased the test so it would actually try to delete something

alf239 committed:
Fixed deleteProperty() method; fixed tests.

alf239 committed:
More uniform property scanning

alf239 committed:
Dropping the oblsolete counter

rpetry committed:
Ticket #3022 - Moved DAO access from CurationApiController to CurationService; misc other fixes

rpetry committed:
Ticket #3022 - Removed superfluous config

rpetry committed:
Ticket #3022 - Addressed code review comments - removed superfluous settings from A2_ONTOLOGY_SEQ sequence

alf239 commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/api/ApiArrayDesign.java:
```diff
@@ -0,0 +1,66 @@
+    private String type;
+//    private Long mappingSoftwareId;
+//    private Map<String, Long> designElements = new HashMap<String, Long>();
+//    private Map<Long, List<Long>> genes = new HashMap<Long, List<Long>>();
```
Could we please get rid of commented-out code? No one will ever be able to find out why it was left here.

rpetry replied:
Done

alf239 commented on atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/SampleDAO.java:
```diff
@@ -0,0 +1,25 @@
+ * User: rpetry
+ * Date: 7/13/11
+ * Time: 3:57 PM
+ * To change this template use File | Settings | File Templates.
```
`use File | Settings | File Templates.`

rpetry replied:
Ammended

alf239 commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Assay.java:
```diff
@@ -214,6 +214,46 @@ public class Assay {
+        }
+
+        return null;
+    }
```
The three methods above are suspiciously similar :)

rpetry replied:
Refactored

alf239 commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/Sample.java:
```diff
@@ -106,6 +103,8 @@ public class Sample {
 
+
+
     public long getSampleID() {
```
it must be `Long` I believe

rpetry replied:
Corrected - also in Assay class

rpetry committed:
Ticket #3022 - more review comments - decoupled CurationService from HTTP transport layer

alf239 commented on atlas-utils/src/main/java/uk/ac/ebi/gxa/utils/GuavaUtil.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * @author Misha Kapushesky
+ */
+public class GuavaUtil {
```
It has nothing to do with Guava actually. `CollectionUtil` perhaps?

rpetry replied:
Renamed to TransformerUtil, as agreed

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,513 @@
+            // TODO: 4ostolop: should we or should we not keep it?
+
+//            log.info("Deleting experiment " + experiment.getAccession() + " in order to update");
+//            experimentDAO.delete(experiment);
```
should we or should we not keep it?

rpetry replied:
As discussed, removed `saveExperiment()` method for now. If `experimentDAO.delete(experiment)` is left commented out, this method will create a new experiment with the same accession but different `experimentid` (wrong). If `experimentDAO.delete(experiment)` is uncommented, the old experiment with same accession willl be replaced with the new one, but also with a different `experimentid`. This is kind of what experiment re-loading currently does, but the new experimentid in the save experiment means that all the processing history for that experiment's accession in Atlas admin API is lost, which is not desirable either.

The 'save experiment' functionality is not currently offered as part of the Curation API. Until the requirements on saving experiments are fully clarified, @alf239 and I made a decision to remove `saveExperiment()` method.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,513 @@
+     * @param response
+     * @throws ResourceNotFoundException if experiment: experimentAccession or assay: assayAccession in that experiment are not found
+     */
+    @Transactional
```
It's enough to make the whole class `@Transactional`, I think. Do we really need non-transactional methods here?

rpetry replied:
Keeping all service setter/getter methods in CurationService keeps separation of concerns clear - the service deals with DAO access, the controller looks after HTTP transport. Assuming we agree on the above, annotating @Transactional at the class level would make the getter calls more costly than they need to.

alf239 replied:
Well, I see no connection between separation of concerns and positioning of `@Transactional`. 

As of the costs, you may be right, I'll check.

rpetry committed:
Ticket #3022 - Added @Nonnull annotations

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,470 @@
+     */
+    private <T> void checkIfFound(T entity, String accession) throws ResourceNotFoundException {
+        if (entity == null) {
+            throw new ResourceNotFoundException("No records for " + entity.getClass().getName() + accession);
```
# Null pointer dereference

A `null` pointer is dereferenced here.  This will lead to a `NullPointerException` when the code is executed.

rpetry replied:
Good spot - fixed.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,470 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.ac.ebi.gxa.dao.*;
+import uk.ac.ebi.gxa.web.controller.ResourceNotFoundException;
```
It's still `web.controller`, we cannot use it if we want to keep code clean.

rpetry replied:
Moved out ResourceNotFoundException to atlas-utls

alf239 commented:
@olgamelnichuk, could you please review the Controller part? I'm not good at mapping exceptions to HTTP statuses; e.g. should I update status by hand if I throw an exception?

rpetry committed:
Ticket #3022 - incorporated more core review comments

rpetry committed:
Ticket #3022 - incorporated more core review comments  - renamed GuavaUtil to TransformerUtil

alf239 commented on atlas-dao/src/test/java/uk/ac/ebi/gxa/dao/AtlasDAOTestCase.java:
```diff
@@ -404,18 +402,18 @@ public abstract class AtlasDAOTestCase extends DataSourceBasedDBTestCase {
+        runStatement(conn, "CREATE SEQUENCE A2_ASSAY_SEQ START WITH 10000000");
+        runStatement(conn, "CREATE SEQUENCE A2_ASSAYPV_SEQ START WITH 10000000");
+        runStatement(conn, "CREATE SEQUENCE A2_ASSET_SEQ START WITH 10000000");
+        runStatement(conn, "CREATE SEQUENCE A2_EXPERIMENT_SEQ START WITH 10000000");
```
Actually it doesn't work for the experiments: they're using "proper" IDs (as we needed to have matching NetCDF name long, long ago), and real values are around 10E7. We should update DB snapshot, I believe.

alf239 commented on atlas-dao/src/test/java/uk/ac/ebi/gxa/dao/AtlasDAOTestCase.java:
```diff
@@ -353,14 +353,12 @@ public abstract class AtlasDAOTestCase extends DataSourceBasedDBTestCase {
 
         runStatement(conn,
                 "  CREATE TABLE A2_SAMPLEPVONTOLOGY (\n" +
-                        "    SAMPLEPVONTOLOGYID bigint not null\n" +
```
We need to port the same change to the "real" database as well There's no point adding IDs to an association table. Even if triggers save us from this problem, it's a bad, bad design.

alf239 replied:
Addressed in e32febb60c87f22cb52bf3f79e262da8bd5a0854; will be cherry-picked to develop, too.

alf239 commented on atlas-dao/src/test/java/uk/ac/ebi/gxa/dao/AtlasDAOTestCase.java:
```diff
@@ -353,14 +353,12 @@ public abstract class AtlasDAOTestCase extends DataSourceBasedDBTestCase {
 
         runStatement(conn,
                 "  CREATE TABLE A2_ASSAYPVONTOLOGY (\n" +
-                        "    ASSAYPVONTOLOGYID bigint not null\n" +
```
We need to port the same change to the "real" database as well There's no point adding IDs to an association table. Even if triggers save us from this problem, it's a bad, bad design.

alf239 replied:
Addressed in e32febb; will be cherry-picked to develop, too.

alf239 commented on atlas-dao/src/test/java/uk/ac/ebi/gxa/dao/TestAtlasDAO.java:
```diff
@@ -178,4 +206,30 @@ public class TestAtlasDAO extends AtlasDAOTestCase {
+                ONTOLOGY_NAME,
+                ONTOLOGY_DESCRIPTION,
+                null,
+                ONTOLOGY_VERSION));
```
This line reminds me that most of the `Ontology` and `OntologyTerm` fields are unused. Actually, we only need term's accession. We can of course track ontologies as well, but we don't.

Out of scope, anyway. Created a ticket, http://bar.ebi.ac.uk:8080/trac/ticket/3058

rpetry committed:
Ticket #3022 - incorporated more core review comments  - moved ResourceNotFoundException to atlas-utils

rpetry committed:
Merge branch 'develop' into curation-api

Commit:
dropping "Created by IntelliJ IDEA."

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,469 @@
+    }
+
+    /**
+     * If entity == null, this method sets appropriate response status and then throws ResourceNotFoundException
```
There's no `appropriate response status` any more. Moreover, checking for `null` is a bad practice: if we have an accessin, we're almost sure we know what we're asking for. Not having such an object is an exception.

rpetry replied:
Removed the outdated comment. Not sure I understand the bad practice bit - having null here means that we could not find the object used specified, not that we didn't have the object in the first place. The user may know what object they're looking for, but but may mistype it.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,469 @@
+                            apiOntologyTerm.getOntology().getName(),
+                            apiOntologyTerm.getOntology().getDescription(),
+                            apiOntologyTerm.getOntology().getSourceUri(),
+                            apiOntologyTerm.getOntology().getVersion()));
```
We have the same code repeated 30 lines above. Can't we just create a method?

alf239 replied:
On a second though, we already have a method, `private OntologyTerm getOrCreateOntologyTerm(@Nonnull ApiOntologyTerm apiOntologyTerm)`. starting line 411. Get rid of unnecessary code duplication please.

rpetry replied:
Done.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,469 @@
+                apiOntologyTerm.getOntology().getName(),
+                apiOntologyTerm.getOntology().getDescription(),
+                apiOntologyTerm.getOntology().getSourceUri(),
+                apiOntologyTerm.getOntology().getVersion());
```
We effectively duplicate the above method here, which isn't a good idea. Also, please consider the [Law of Demeter](http://en.wikipedia.org/wiki/Law_of_Demeter): taking `apiOntologyTerm.getOntology()` apart is akin to commanding dog's legs.

rpetry replied:
Modified, as we discussed, to force the user to create a new ontology before assigning ontology terms to it.

olgamelnichuk commented on atlas-dao/pom.xml:
```diff
@@ -94,6 +94,11 @@
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-web</artifactId>
+            <version>${spring.version}</version>
+        </dependency>
```
Why do we need spring-web in atlas-dao module?

rpetry replied:
Have removed - must have been a remnant of some earlier attempts - thanks for spotting it.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/api/CurationApiController.java:
```diff
@@ -0,0 +1,220 @@
+                                       HttpServletResponse response) throws ResourceNotFoundException {
+        try {
+            ApiExperiment apiExperiment = curationService.getExperiment(experimentAccession);
+            response.setStatus(HttpServletResponse.SC_FOUND);
```
According to the javadoc,

``` java
    /**
     * Status code (302) indicating that the resource reside
     * temporarily under a different URI. Since the redirection might
     * be altered on occasion, the client should continue to use the
     * Request-URI for future requests.(HTTP/1.1) To represent the
     * status code (302), it is recommended to use this variable.
     */
    public static final int SC_FOUND = 302;
```

Why do we set this status here?

rpetry replied:
Replaced it with SC_OK - am sure that's what Misha intended to use.

olgamelnichuk replied:
Actually SC_OK is set by default for all responses finished successfully

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/api/CurationApiController.java:
```diff
@@ -0,0 +1,220 @@
+            response.setStatus(HttpServletResponse.SC_FOUND);
+            return apiExperiment;
+        } catch (ResourceNotFoundException e) {
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
```
We already have exception handler in the AtlasViewController:

``` java
    @ExceptionHandler(ResourceNotFoundException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ModelAndView handleError(ResourceNotFoundException e) {
        log.info(e.getMessage());
        return errorDetails(e);
    }
```

rpetry replied:
Thanks—wasn't aware of that—have removed exception handling from `CurationApiController`.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/api/CurationApiController.java:
```diff
@@ -0,0 +1,220 @@
+    public void putExperiment(@RequestBody final ApiExperiment apiExperiment,
+                              HttpServletResponse response) throws ResourceNotFoundException {
+        curationService.saveExperiment(apiExperiment);
+        response.setStatus(HttpServletResponse.SC_CREATED);
```
Please use spring annotation to set the response status:

``` java
@ResponseStatus(HttpStatus.CREATED)
```

rpetry replied:
Done.

rpetry committed:
Ticket #3022 -  ApiSampleProperty and ApiAssayProperty were identical - replaced them both with ApiProperty class. Added Curation API methods to retrieve all properties and all values for a given property.

rpetry committed:
Ticket #3022 -  code review - modified to prevent ontology terms being assigned to non-existing ontology - the user needs to create that ontology first

rpetry committed:
Ticket #3022 -  fixed a bug in deleting assay properties I inadvertently introduced earlier in

rpetry committed:
Ticket #3022 -  More core review comments  - revised HTTP response codes and exception handling in CurationApiController; removed spurious entry spring-web entry from atlas-dao/pom.xml

olgamelnichuk commented:
I don't see any unit tests for db service and trasformation utilities.

rpetry committed:
Ticket #3022 -  Fixed an ExperimentDAO method, add some new ones to AssayDAO, SampleDAO and created junit tests to test the changes

rpetry committed:
Ticket #3022 -  Work in progress on replacing one property value with another in all assays/samples, and on creating junit tests for CurationService methods.

rpetry committed:
Ticket #3022 -  Removed ASSAYSAMPLEID to keep hibernate happy

rpetry committed:
Ticket #3022 -  Various changes to allow objects to be modified and/or saved successfully in junit tests

rpetry committed:
Ticket #3022 -  Added new methods to Curation API and tests for them; removed saveExperiment() for the time being

rpetry committed:
Ticket #3022 -  Corrected property names to lower case

alf239 commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/ArrayDesign.java:
```diff
@@ -36,13 +36,13 @@ public class ArrayDesign {
     private String type;
     @Column(name = "MAPPINGSWID")
-    private long mappingSoftwareId;
+    private Long mappingSoftwareId;
```
The getter will throw a hard-to-diagnose NPE if called when `mappingSoftwareId  == null`

rpetry replied:
Made getMappingSoftwareId() return Long

rpetry committed:
Ticket #3022 - Changed usage - old: to represent just property name; new: to represent Assay or Sample property

alf239 commented on atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/ArrayDesign.java:
```diff
@@ -36,13 +36,13 @@ public class ArrayDesign {
     @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "arrayDesignSeq")
     @SequenceGenerator(name = "arrayDesignSeq", sequenceName = "A2_ARRAYDESIGN_SEQ")
-    private long arrayDesignID;
+    private Long arrayDesignID;
```
The getter will throw a hard-to-diagnose NPE if called when `arrayDesignID == null`

rpetry replied:
Made getArrayDesignID() return Long

rpetry committed:
Replacing with ApiProperty

alf239 commented on sql/Schema/Tables.sql:
```diff
@@ -824,14 +824,13 @@ ALTER TRIGGER A2_SAMPLE_INSERT ENABLE;
-    "ASSAYSAMPLEID" NUMBER(22,0) CONSTRAINT NN_ASSAYSAMPLE_ID NOT NULL
-  , "ASSAYID" NUMBER(22,0) CONSTRAINT NN_ASSAYSAMPLE_ASSAYID NOT NULL
+   "ASSAYID" NUMBER(22,0) CONSTRAINT NN_ASSAYSAMPLE_ASSAYID NOT NULL
   , "SAMPLEID" NUMBER(22,0) CONSTRAINT NN_ASSAYSAMPLE_SAMPLEID NOT NULL) ;
```
Since we never rebuild the curators' instance from scratch, an update script would be helpful. Please refer to https://github.com/gxa/gxa/blob/develop/sql/updates/2.0.8-2.0.9/20110811-1-proper_associations.sql for an example.

rpetry replied:
Added DB update files into 2.0.9-2.0.10 dir.

alf239 commented on sql/Schema/Tables.sql:
```diff
@@ -861,26 +860,6 @@ ALTER TRIGGER A2_SAMPLE_INSERT ENABLE;
-for each row
-begin
-if(:new.AssaySampleID is null) then
-select A2_AssaySample_seq.nextval into :new.AssaySampleID from dual;
```
We're removing the usage of `A2_AssaySample_seq`, bur not `A2_AssaySample_seq` itself, why?

rpetry replied:
You're right - we should - I missed that due to different casing in the creation statement - I've removed it now.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,538 @@
+        Property property = propertyDAO.getByName(propertyName);
+        checkIfFound(property, Property.class, propertyName);
+        PropertyValue propValue = propertyValueDAO.find(property, propertyValue);
+        checkIfFound(propValue, PropertyValue.class, propertyValue);
```
Use checked exceptions instead, please. You will throw an exception anyway, it's just that this exceptional situation pollutes the code: we return `null` in DAO, making check for that, and then we're trying not to forget this `null` check in the clients. Exceptions are technically impossible to forget, and they structure the code better.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,538 @@
+        for (Sample sample : samples) {
+            sample.deleteProperty(propValue);
+            sampleDAO.save(sample);
+        }
```
Can't we just enable cascading on delete?

rpetry replied:
Sorry - missed that - cascading was already enabled - I didn't need to cycle through assays/samples here - thanks

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,538 @@
+
+        // Now delete propertyName:propertyValue itself
+        propertyDAO.delete(property, propValue);
+        propertyDAO.save(property);
```
Two calls to one object in a row almost always mean you're doing something that object would've pretty much able to do on its own.

rpetry replied:
Yep - moved the second call to  propertyDAO.delete()

Commit:
`allocationSize = 1` in order to disable Hibernate's hilo approach (which is nice, but not ready for our DB yet)

Commit:
Fixing `ATLASMGR` package and rebuilding the sequences

 - `ATLASMGR` was apparently out-of-date in curators' instance
 - The previously fixed bug with `hilo` approach could get sequences out of sync with the tables, so it won't hurt running `RebuildSequences()` just to be sure (it's idempotent, anyway)

Commit:
The actual call was missing

rpetry committed:
Ticket #3022 -  More review feedback incorporated

rpetry committed:
Database update files

rpetry committed:
Ticket #3022 -  more code review feedback - Modified getByName() methods in DAO classes to return DAOException (nstead of returning null) when entity matching the specified name was not found.

rpetry committed:
Ticket #3022 -  more code review feedback - Modified getByName() methods in DAO classes to return DAOException (nstead of returning null) when entity matching the specified name was not found - a remainder I missed in my previous commit

rpetry committed:
Ticket #3022 -  Added missing FK constraints to A2_ASSAYPVONTOLOGY and A2_SAMPLEPVONTOLOGY

rpetry committed:
Ticket #3022 -  Unified handling of new Ontologies when an ontology term is updated vs created - in both cases a new Ontology is created automatically.

rpetry committed:
Ticket #3022 -  Clear hibernate session cache to force re-load of objects updated via DB-only foreign key constraint 'ON DELETE CASCADE' operations.

rpetry committed:
Merge branch 'develop' into curation-api

Conflicts:
	atlas-analytics/src/main/java/uk/ac/ebi/gxa/analytics/generator/service/ExperimentAnalyticsGeneratorService.java
	atlas-loader/src/main/java/uk/ac/ebi/gxa/loader/service/AtlasNetCDFUpdaterService.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java

olgamelnichuk commented:
Why we start to use microarray/atlas in the package name (e.g. uk/ac/ebi/microarray/atlas/api/)?

alf239 commented on atlas-loader/src/main/java/uk/ac/ebi/gxa/loader/dao/LoaderDAO.java:
```diff
@@ -36,11 +43,22 @@ public class LoaderDAO {
+            // do nothing - valid situation
+        }
         if (propertyValue == null) {
             propertyValueDAO.save(propertyValue = new PropertyValue(null, property, value));
```
Why cannot we do it "catch" clause and keep login in one place?

rpetry replied:
I guess we can - it makes the code simpler. done.

alf239 commented on atlas-loader/src/main/java/uk/ac/ebi/gxa/loader/dao/LoaderDAO.java:
```diff
@@ -27,7 +28,13 @@ public class LoaderDAO {
+        }
+
         if (organism == null) {
             organismDAO.save(organism = new Organism(null, name));
```
Why cannot we do it "catch" clause and keep login in one place?

rpetry replied:
Done

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,533 @@
+        }
+
+        if (ontology == null) {
+            ontology = getOrCreateOntology(apiOntology);
```
Why cannot we do it "catch" clause and keep login in one place?

rpetry replied:
Done

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,533 @@
+            ontology.setDescription(apiOntology.getDescription());
+            ontology.setName(apiOntology.getName());
+            ontology.setVersion(apiOntology.getVersion());
+            ontology.setSourceUri(apiOntology.getSourceUri());
```
As far as I can see, no test cover this branch. Do we really need it?

rpetry replied:
testPutOntology() in TestCurationService, no?

alf239 replied:
It only creates the ontology, never updates it.

rpetry replied:
Ah, right, added test

alf239 commented:
As far as I can see, all the `// do nothing - valid situation` situations can and should be handled in the `catch` clause, there's no need to check for `null` after `try`...`catch`.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,533 @@
+                ontologyTerm.setDescription(apiOntologyTerm.getDescription());
+                Ontology ontology = getOrCreateOntology(apiOntologyTerm.getOntology());
+                ontologyTerm.setOntology(ontology);
+                ontologyTerm.setTerm(apiOntologyTerm.getTerm());
```
This branch is not covered; do we need it?

rpetry replied:
testPutOntologyTerms() in TestCurationService does

alf239 replied:
No it doesn't, put a breakpoint and check it out.

rpetry replied:
Tests now cover creation and updating of ontology terms  - I've put a breakpoint on both sections and checked that they are tested.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,533 @@
+                List<OntologyTerm> terms = oldSampleProperty.getTerms();
+                sample.deleteProperty(oldPropertyValue);
+                if (newSampleProperty != null) {
+                    terms.addAll(newSampleProperty.getTerms());
```
This branch is not covered

rpetry replied:
See testReplacePropertyValueInSamples() in TestCurationService

alf239 replied:
My debugger doesn't think so.

rpetry replied:
Added tests to cover this branch

Commit:
Code style

Commit:
Dropping ID from equals and hashCode

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,533 @@
+
+            List<OntologyTerm> terms = Lists.newArrayList();
+            for (ApiOntologyTerm apiOntologyTerm : apiAssayProperty.getTerms()) {
+                terms.add(getOrCreateOntologyTerm(apiOntologyTerm));
```
This cycle is never tested, as `apiAssayProperty.getTerms()` is always empty

Commit:
Eliminated a bit of duplicated code

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,532 @@
+ */
+@Service
+public class CurationService {
+    private final Logger log = LoggerFactory.getLogger(getClass());
```
We never log anything here

Commit:
Unnecessary variables

Commit:
Never ever return dummy values if the method is not supposed to be used

First of all, it's a good sign of bad design.
Secondly, if something unexpected happens, it must not get unnoticed.

Commit:
Keep abstract methods abstract

we referred to this method in the javadocs saying it's abstract; and it should be, "name" is not something too common here, hence no need to pull it to the parent class

Commit:
UnsupportedOperationException, mea culpa

rpetry committed:
Ticket #3022 - Core review comments - moved creation of objects not found by DAO into catch(DAOException) {} clause

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,526 @@
+            }
+
+            if (ontologyTerm == null) { // ontology term already exists - update it
+                ontologyTerm.setAccession(apiOntologyTerm.getAccession());
```
NPE?

rpetry replied:
Thanks - eliminated now.

alf239 commented on atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:
```diff
@@ -0,0 +1,526 @@
+        }
+
+        if (ontology == null) { // ontology already exists - update it
+            ontology.setDescription(apiOntology.getDescription());
```
NPE?

rpetry replied:
Thanks - eliminated now.

rpetry committed:
Merge branch 'develop' into curation-api

rpetry committed:
Merge branch 'develop' into curation-api

Conflicts:
	atlas-dao/src/test/java/uk/ac/ebi/gxa/dao/TestAtlasDAO.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java

Ticket #3022 - more junit tests

rpetry committed:
Ticket #3022 - Eliminated sampleid from equals()/hashCode() - to reflect an equivalent change in Assay class

rpetry committed:
Ticket #3022 - Fixed AD accession to make TestAtlasDAO pass test

rpetry committed:
Ticket #3022 - Corrected RequestMethod for removing property values from from PUT to DELETE

Commit:
Merge branch 'develop' into curation-api

Conflicts:
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java

Commit:
Apparently missing changes from the merge

Commit:
code style

Commit:
Simplified the close() method

Commit:
Reverting unnecessary Experiment retrieval

Commit:
Removing now-unused ExperimentDAO

Commit:
Removing now-unused DAOException's catch clause

alf239 commented on atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/hibernate/DAOException.java:
```diff
@@ -0,0 +1,13 @@
+package uk.ac.ebi.gxa.dao.hibernate;
+
+/**
+ * Thrown when DAO's getByName() method returns no results
```
The class name is _way_ too generic to restrict its usage to _one_ method.

alf239 committed:
More specific exceptions, more idiomatic Java

alf239 committed:
DRY - Don’t repeat yourself

This is probably the single most fundamental tenet in programming is to avoid repetition. Many programming constructs exist solely for that purpose (e.g. loops, functions, classes, and more). As soon as you start repeating yourself (e.g. a long expression, a series of statements, same concept) create a new abstraction. http://en.wikipedia.org/wiki/Don%27t_repeat_yourself

alf239 committed:
From start/end session to declarative transactions

alf239 committed:
Prefer specialized DAO over AtlasDAO

alf239 committed:
Dropping now-unnecessary dependencies

alf239 commented:
Minimised the added code; dropped repeating functionality in the old code where necessary.

@olgamelnichuk, @rpetry — could you please make a final review before the merge?

alf239 committed:
Merge branch 'develop' into curation-api

Conflicts:
	atlas-web/src/main/java/ae3/model/ExperimentalData.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/api/ApiQueryRequestHandler.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentDesignViewController.java
	atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java

alf239 committed:
Unused @Before and @After (inherited members are exactly the same)
